fix: fix bazel-out path for `rules_docker-0.23.0`

sashamor opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Include if applicable: -->
Fixes: #7228 <!-- tracking issues that this PR will close -->
**Related**: https://github.com/bazelbuild/rules_docker/pull/1963
**Merge before/after**: _Dependent or prerequisite PRs_

**Description**
<!-- Describe your changes here. The more detail, the easier the review! -->
With the release of https://github.com/bazelbuild/rules_docker/releases/tag/v0.23.0, the output directory for bazel targets is calculated differently and is no longer a static path.

For example, a `container_image` target such as:
```
//foo/bar/server:server_img.tar
```

Previously was stored at:
```
bazel-out/k8-fastbuild/bin/foo/bar/server/server_img.tar
```

Internally, `skaffold` calculated this path by running `bazel info bazel-bin`, which returns the start of the path (e.g. `/home/...../execroot/__main__/bazel-out/k8-fastbuild/bin`), then appended the name of the rule with `:` replaced with `/` (e.g. `foo/bar/server/server_img.tar`). This results in the correct path as above.

However, with `rules_docker-0.23.0`, the output is now stored at a path like:
```
bazel-out/k8-fastbuild-ST-4a519fd6d3e4/bin/foo/bar/server/server_img.tar
```

So `bazel info` no longer returns the correct path prefix.

Using `bazel cquery` with a similar method to that described in https://github.com/bazelbuild/rules_pkg/tree/main/examples/where_is_my_output, as recommended in https://github.com/bazelbuild/rules_docker/pull/1963#issuecomment-1081197607, we can query Bazel for the actual output path to the target. Docker `.tar` output files are available in `cquery` as targets with a `.files` attribute, which contains the list of output files (there's only one), and each file has a `.path` attribute that gives the path to the file, which we can use internally in Skaffold. This is also backwards compatible with older versions of https://github.com/bazelbuild/rules_docker.

For example, the command:
```
bazel cquery //foo/bar/server:server_img.tar --output starlark --starlark:expr="target.files.to_list()[0].path"
```
May return:
```
bazel-out/k8-fastbuild/bin/foo/bar/server/server_img.tar
``` 
And with `rules_docker-0.23.0`, this may return something like:
```
bazel-out/k8-fastbuild-ST-4a519fd6d3e4/bin/foo/bar/server/server_img.tar
```
Which is the correct path to the target.

<!--
Please be sure your PR includes unit tests - we don't merge code that brings down test coverage! 
Integration tests are sometimes an appropriate substitute. 
-->

MarlonGamez commented:
hi @sashamor , Thanks for opening this change! It looks like you'll have to go in and update `TestBuildBazel` unit test as well, but other than that these changes look good :)

sashamor commented:
Thanks, fixed :) Also changed the arguments to no longer use nested quotes.

Tested locally with https://github.com/bazelbuild/rules_docker v0.23.0 and v0.22.0 and can confirm they both find the output tar correctly as expected.

Anthony-Bible commented:
I wanted to add another data point, I had a similar problem and it worked when running with your patch.

MarlonGamez commented:
@sashamor , thanks for making the change! I have a fix open for the github action failure we're seeing 
https://github.com/GoogleContainerTools/skaffold/pull/7263
Once this is merged, could you rebase this PR to retrigger the tests? I think it should be fixed then

sashamor commented:
Thanks for rebasing and merging :) Woohoo

morozov commented on pkg/skaffold/build/bazel/build.go:
```diff
@@ -80,12 +79,11 @@ func (b *Builder) buildTar(ctx context.Context, out io.Writer, workspace string,
 	}
 
-	bazelBin, err := bazelBin(ctx, workspace, a)
+	tarPath, err := bazelTarPath(ctx, workspace, a)
```
It looks like this breaks the scenarios where the artifact has a non-default context. The `bazel cquery` runs in the directory corresponding to the context and returns that tar path relative to that context but `docker.Push()` still runs in the current working directory and is now supplied with a path relative to the context.

Take the following example. There is a scaffold file in projects/acme/some-projects:
```yaml
apiVersion: skaffold/v2beta26
kind: Config
build:
  artifacts:
  - image: quay.io/acme/some-project
    context: ../../..
```
If one does:
```
cd projects/acme/some-projects && skaffold build
```
It will no longer work. It will fail with an error like:
```
build [quay.io/sugarcrm/blackhole] failed: reading image "bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project/docker.tar": open bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project//docker.tar: no such file or directory
```

morozov commented on pkg/skaffold/build/bazel/build.go:
```diff
@@ -80,12 +79,11 @@ func (b *Builder) buildTar(ctx context.Context, out io.Writer, workspace string,
+		return "", fmt.Errorf("getting bazel tar path: %w", err)
 	}
 
-	tarPath := filepath.Join(bazelBin, buildTarPath(a.BuildTarget))
```
It looks like this breaks the scenarios where the artifact has a non-default context. The `bazel cquery` runs in the directory corresponding to the context and returns the tar path relative to that context but `docker.Push()` still runs in the current working directory and is now supplied with a path relative to the context.

Take the following example. There is a skaffold file in projects/acme/some-projects:
```yaml
apiVersion: skaffold/v2beta26
kind: Config
build:
  artifacts:
  - image: quay.io/acme/some-project
    context: ../../..
```
If one does:
```
cd projects/acme/some-projects && skaffold build
```
It will no longer work. It will fail with an error like:
```
build [quay.io/sugarcrm/blackhole] failed: reading image "bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project/docker.tar": open bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project//docker.tar: no such file or directory
```

morozov replied:
It looks like this breaks the scenarios where the artifact has a non-default context and the `skaffold` command is executed from within a path different than the repository root.

The `bazel cquery` runs in the directory corresponding to the context and returns the tar path relative to that context but `docker.Push()` still runs in the current working directory and is now supplied with a path relative to the context, not relative to the current working directory.

Take the following example. There is a skaffold file in projects/acme/some-projects:
```yaml
apiVersion: skaffold/v2beta26
kind: Config
build:
  artifacts:
  - image: quay.io/acme/some-project
    context: ../../..
```
If one does:
```
cd projects/acme/some-projects && skaffold build
```
It will no longer work. It will fail with an error like:
```
build [quay.io/acme/some-project] failed: reading image "bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project/docker.tar": open bazel-out/darwin-fastbuild-ST-4a519fd6d3e4/bin/projects/some-project//docker.tar: no such file or directory
```
